sycl/doc/design/SYCLBINDesign.md

```diff
@@ -0,0 +1,272 @@
+# SYCLBIN - A format for separately compiled SYCL device code
+
+This design document details the SYCLBIN binary format used for storing SYCL
+device binaries to be loaded dynamically by the SYCL runtime. It also details
+how the toolchain produces, links and packages these binaries, as well as how
+the SYCL runtime library handles files of this format.
+
+
+## SYCLBIN binary format
+
+The files produced by the new compilation path will follow the format described
+in this section. The intention of defining a new format for these is to give
+the DPC++ implementation an extendable and lightweight wrapper around the
+multiple modules and corresponding metadata captured in the SYCLBIN file.
+The content of the SYCLBIN may be contained as an entry in the offloading binary
+format produced by the clang-offload-packager, as described in
+[ClangOffloadPackager.rst](https://github.com/intel/llvm/blob/sycl/clang/docs/ClangOffloadPackager.rst).
+
+The format of these files consist of a [file header](#file-header), containing
+general information about the file, followed by three lists of headers: The
+[abstract module header](#abstract-module-header) list, the
+[IR module header](#ir-module-header) list and
+[native device code image header](#native-device-code-image-header) list,
+containing information about the [abstract modules](#abstract-module),
+[IR modules](#ir-module) and
+[native device code images](#native-device-code-image) respectively.
+
+|                                                                       |
+| --------------------------------------------------------------------- |
+| [File header](#file-header)                                           |
+| [Abstract module header](#abstract-module-header) 1                   |
+| ...                                                                   |
+| [Abstract module header](#abstract-module-header) N                   |
+| [IR module header](#ir-module-header) 1                               |
+| ...                                                                   |
+| [IR module header](#ir-module-header) M                               |
+| [Native device code image header](#native-device-code-image-header) 1 |
+| ...                                                                   |
+| [Native device code image header](#native-device-code-image-header) L |
+| Metadata byte table                                                   |
+| Binary byte table                                                     |
+
+The headers and each byte table are all aligned to 8 bytes. The fields in the
+headers use C/C++ type notation, including the fixed-size integer types defined
+in the `<cstdint>` header, and will have the same size and alignment.
```

Comment:
Probably should state that the numbers are little endian.  Just in case someone decides to port SYCL to a 1990's era Macintosh. ;-)